### PR TITLE
Start Building ARM VMs

### PR DESCRIPTION
**Pull request by boumenot**
This is the _final_ PR to realize support for ARM, and by _final_ I really mean that is will work end-to-end.  I've logged issues to cover the defects found in code review as well as necessary enhancements to make this a **real** solution.

Part of the changes needed for #94

### CODE REVIEW

File: packer/builder/azure/arm/artifact.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Open Technologies, Inc.
+// All Rights Reserved.
+// Licensed under the Apache License, Version 2.0.
+// See License.txt in the project root for license information.
```
Comment (paulmey) [inline packer/builder/azure/arm/artifact.go]: Change license

File: packer/builder/azure/arm/builder.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	return warnings, errs
+}
+
+func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packer.Artifact, error) {
+	ui.Say("Preparing builder ...")
+	ui.Message("Creating Azure Resource Manager (ARM) client ...")
+
+	// TODO(chrboum): figure out how to implement a test case to validate these properties
```
Comment (paulmey) [inline packer/builder/azure/arm/builder.go]: Create issue, remove TODO

File: packer/builder/azure/arm/builder.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return warnings, errs
+}
+
+func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packer.Artifact, error) {
+	ui.Say("Preparing builder ...")
+	ui.Message("Creating Azure Resource Manager (ARM) client ...")
```
Comment (paulmey) [inline packer/builder/azure/arm/builder.go]: neater to move this closer to the actual creation call (of either the principal or the actual client)

File: packer/builder/azure/arm/builder.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		&communicator.StepConnectSSH{
+			Config:    &b.config.Comm,
+			Host:      lin.SSHHost,
+			SSHConfig: lin.SSHConfig(b.config.UserName),
+		},
+		&common.StepProvision{},
+		&lin.StepGeneralizeOS{
+			Command: "sudo /usr/sbin/waagent -force -deprovision+user && export HISTSIZE=0 && sync",
+		},
+		NewStepQueryVM(azureClient, ui),
```
Comment (paulmey) [inline packer/builder/azure/arm/builder.go]: yeah, as a reader, I'm not sure what we're doing here... (querying... what or why?)

File: packer/plugin/packer-builder-azure-arm/main_test.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+// Copyright (c) Microsoft Open Technologies, Inc.
+// All Rights Reserved.
+// Licensed under the Apache License, Version 2.0.
+// See License.txt in the project root for license information.
```
Comment (paulmey) [inline packer/plugin/packer-builder-azure-arm/main_test.go]: IMHO, we can leave this file out. It's cheating, in that it suggests we have tests while we don't. :-)
If you still want to keep it (after being called a cheater), change the license.

:stuck_out_tongue_winking_eye:

File: packer/plugin/packer-builder-azure-arm/main.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft Open Technologies, Inc.
+// All Rights Reserved.
+// Licensed under the Apache License, Version 2.0.
+// See License.txt in the project root for license information.
```
Comment (paulmey) [inline packer/plugin/packer-builder-azure-arm/main.go]: Change license

**Comment by paulmey** [comment]
How do you feel with the `-arm-` in the ARM builder vs. the `-sm-` in the ASM post-processor?

File: packer/builder/azure/arm/artifact.go (hunk 1 above)
Comment (boumenot) [inline packer/builder/azure/arm/artifact.go]: Gah!  I should have caught that myself.

File: packer/plugin/packer-builder-azure-arm/main_test.go (hunk 5 above)
Comment (boumenot) [inline packer/plugin/packer-builder-azure-arm/main_test.go]: :smile:  I'll delete it.

File: packer/builder/azure/arm/builder.go (hunk 2 above)
Comment (boumenot) [inline packer/builder/azure/arm/builder.go]: Issue #170.

File: packer/builder/azure/arm/builder.go (hunk 4 above)
Comment (boumenot) [inline packer/builder/azure/arm/builder.go]: This is covered by #156.

**Comment by boumenot** [comment]
I don't like that we have to surface this distinction to the users or label the tools in such a way, but it isn't clear to me how to erase that difference given the state of the world.  Some tools do not have the -sm- because -arm- came afterwards and needed to distinguish itself.  Should we go back and correct that, or leave it as is?

So, I fine with the labels, and we can eventually remove them I would vote it favor of that.

**Pull request merged**